### ISSUE TITLE
Conditional PackageReference version splitting analyzer presence fails

### ISSUE THREAD

**Issue by rainersigwald** [dotnet/sdk#18148]
If a project references two different versions of the same package in two different TargetFrameworks, and the package has added or removed a Roslyn analyzer, the analyzer will be added from _both_ versions of the package, even the one where it's not present.

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>net6.0;net472</TargetFrameworks>
    <RootNamespace>two_systemtextjson</RootNamespace>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Text.Json" Version="4.7.0" Condition=" '$(TargetFramework)' == 'net472' " />
    <PackageReference Include="System.Text.Json" Version="6.0.0-preview.4.21253.7" Condition=" '$(TargetFramework)' == 'net6.0' " />
  </ItemGroup>

</Project>
```

```sh-session
two-systemtextjson on master .NET v6.0.100-preview.5.21302.13 🎯 net6.0;net472 
❯ dotnet build  
Microsoft (R) Build Engine version 17.0.0-preview-21302-02+018bed83d for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored S:\work\two-systemtextjson\two-systemtextjson.csproj (in 135 ms).
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
CSC : error CS0006: Metadata file 'C:\Users\raines\.nuget\packages\system.text.json\4.7.0\analyzers\dotnet\cs\System.Text.Json.SourceGeneration.dll' could not be found [S:\work\two-systemtextjson\two-systemtextjson.csproj]
  two-systemtextjson -> S:\work\two-systemtextjson\bin\Debug\net6.0\two-systemtextjson.dll

Build FAILED.

CSC : error CS0006: Metadata file 'C:\Users\raines\.nuget\packages\system.text.json\4.7.0\analyzers\dotnet\cs\System.Text.Json.SourceGeneration.dll' could not be found [S:\work\two-systemtextjson\two-systemtextjson.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.20
```

Because 4.7.0 of that package didn't have an analyzer/source generator, the file isn't found in that package version. The reference is only there because it was found in the other version.

**Comment by Eilon** [comment]
We just ran into this with the newly-introduced analyzer in Microsoft.Extensions.Logging.Abstractions 6.0.0-preview.5.21301.5. We found a fairly easy workaround for this, as long as you don't actually need the analyzer in the project:

```csproj
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFrameworks>netstandard2.1;net6.0-android</TargetFrameworks>
  </PropertyGroup>
  <ItemGroup Condition="'$(TargetFramework.Contains(net6.0))' == 'true'">
    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="6.0.0-preview.5.21301.5" />
  </ItemGroup>
  <ItemGroup Condition="'$(TargetFramework.Contains(net6.0))' != 'true'">
    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="5.0.0" />
  </ItemGroup>

  <!-- Workaround for this bug (replace the analyzer name with the one you need to exclude (filename only, no extension) -->
  <Target Name="RemoveLoggingAnalyzer" BeforeTargets="CoreCompile">
    <ItemGroup>
      <Analyzer Remove="@(Analyzer)" Condition="%(FileName) == 'Microsoft.Extensions.Logging.Generators'" />
    </ItemGroup>
  </Target>
</Project>
```

**Comment by dsplaisted** [comment]
This looks like a symptom of the following: https://github.com/NuGet/Home/issues/6279

FYI @JonDouglas

**Comment by JonDouglas** [comment]
Can anyone on this thread help us understand if this issue is blocked by the NuGet implementation of https://github.com/NuGet/Home/issues/6279? Just to help us understand if this is a symptom or another issue entirely.

**Comment by ericstj** [comment]
I think NuGet/Home#6279 if fixed would fix this issue since it would let the SDK delete the code with this bug (and replace it with something else).

In lieu of that, I believe this can be fixed in the SDK.

Bug is here:
https://github.com/dotnet/sdk/blob/529c048b7d7d4b04974fa40ce4073693eb8a0916/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs#L836-L855


This enumerates every library referenced by the all TargetFrameworks of the library.  If any library contains an analyzer (on any targetframework) it then uses the package for the current _runtimeTarget and creates an item as if the asset came from that package.

This is broken in a couple ways:
1. using the runtime target for compile time assets (analyzers) is wrong.  Compile target should be used.
2. using all packages as the input is wrong.  Need to use compile target as the input, all packages should serve as the lookup (to get file list and lookup analyzer).

This algorithm should instead be:
1. Enumerate each package referenced in the _compileTarget.
2. Lookup that package/version in the _lockFile.Libraries.
3. Enumerate files.
4. If a file is an applicable analyzer, raise that as an analyzer.

**Comment by bruno-garcia** [comment]
> I think NuGet/Home#6279 if fixed would fix this issue since it would let the SDK delete the code with this bug (and replace it with something else).

@ericstj The reply from @zkat (https://github.com/NuGet/Home/issues/6279#issuecomment-870000004) on that issues was that:

> Our understanding is that dotnet/sdk#18148 is not strictly blocked by this issue, as it can be implemented without NuGet's help. If this is not the case, please update us (and @ us). /cc @bruno-garcia

Bringing this to light as [this completely blocks our ability to add a .NET 6 target](https://github.com/getsentry/sentry-dotnet/pull/939/files#diff-3c95bfba54399ea2c9677a4c600328a38fc95ec8dc9398e01356ea0481a52ef8R16-R34) to our package target `Microsoft.Extensions.Logging` so pretty critical on our end:

https://github.com/getsentry/sentry-dotnet/blob/7d008a9fc437d86d6eeff33300f2ba0160c74a9c/src/Sentry.Extensions.Logging/Sentry.Extensions.Logging.csproj#L16-L29

That's unless we move all the TFMs to depend on the net6 binaries which can cause some diamond dependency and/or other breakages and we'd like to avoid. For reference [this package is relatively popular with 6M downloads](https://nugt.net/s/po8rD2n).

**Comment by Eilon** [comment]
@bruno-garcia please check the workaround I posted above: https://github.com/dotnet/sdk/issues/18148#issuecomment-860957180

I know it's not ideal, but it should un-block you for now.

**Comment by TanvirArjel** [comment]
@Eilon, I have done your workaround but still getting the following error:

![Captureee](https://user-images.githubusercontent.com/14342773/125338505-18654900-e372-11eb-8545-40ca92992799.PNG)

Here is my `.csproj`

```C#
<ItemGroup>
  <PackageReference Include="Blazored.LocalStorage" Version="4.1.2" />
  <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="5.0.*" />
  <PackageReference Include="Microsoft.AspNetCore.Components.Authorization" Version="5.0.*" />
  <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.333">
	  <PrivateAssets>all</PrivateAssets>
	  <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
  </PackageReference>
  <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="6.11.1" />
</ItemGroup>

<ItemGroup Condition="'$(TargetFramework.Contains(net6.0))' == 'true'">
  <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="6.0.0-preview.5.21301.17" />
  <PackageReference Include="Microsoft.AspNetCore.Components.Authorization" Version="6.0.0-preview.5.21301.17" />
</ItemGroup>

<!-- Workaround for this bug (replace the analyzer name with the one you need to exclude (filename only, no extension) -->
<Target Name="RemoveLoggingAnalyzer" BeforeTargets="CoreCompile">
  <ItemGroup>
	  <Analyzer Remove="@(Analyzer)" Condition="%(FileName) == 'Microsoft.Extensions.Logging.Generators'" />
  </ItemGroup>
</Target>
```

**Comment by Eilon** [comment]
@TanvirArjel hmm that's odd, I would think it should have worked because it seems that it's the same issue I was having. Can you try deleting the bin and obj folders or otherwise clearing the git repo and try again? And double check that you made this change to the right csproj (`TanvirArjel.Blazor.csproj`)?

**Comment by ericstj** [comment]
I don't think it's practical to ask one package to add targets to remove assets from another package.  That's not a reasonable solution.

I guess we could "give up" on using the `analyzers` nuget convention entirely and instead just define them in props/targets. 
 NuGet handles props/targets better in that it will actually honor Include/Exclude transitively (and also won't have this bug). As a con it would increase the evaluation time of projects consuming the package.  @dsplaisted / @rainersigwald / @jaredpar / @chsienki / @KathleenDollard what are your thoughts?

**Comment by dsplaisted** [comment]
I went ahead and sent a PR to fix this: #18877

Thanks to @ericst for identifying where and what the fix was.

I still would really like to see NuGet fix this class of issue by treating analyzers as a proper asset type, though I recognize that's a lot more complicated.

**Comment by TanvirArjel** [comment]
@Eilon Tried everything you said but did not work.

**Comment by MarkStega** [comment]
@Eilon  I have the same issue as @TanvirArjel -- I added the suggested content to my csproj but the problem persists. DN6P6 & VS2022P2

**Comment by dsplaisted** [comment]
@TanvirArjel @MarkStega We're hoping to have this fixed in .NET 6 Preview 7.  If you are able to provide a [binlog](https://aka.ms/binlog) of where the workaround isn't working, we may be able to figure out why and how to fix it.

**Comment by MarkStega** [comment]
@dsplaisted Here is a binlog of my failing build as well as the pertinent content of the csproj in question
[_Debug_AnyCPU_Build_2021-07-15T09_54_49.2661489-04_00.binlog.zip](https://github.com/dotnet/sdk/files/6823611/_Debug_AnyCPU_Build_2021-07-15T09_54_49.2661489-04_00.binlog.zip)
```
    <!-- Conditionally obtain references for the .NET Framework 6.0 target -->
    <ItemGroup Condition=" '$(TargetFramework)' == 'net6.0' ">
        <PackageReference Include="Microsoft.AspNetCore.Components" Version="6.0.0-preview.5.21301.17" />
        <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="6.0.0-preview.5.21301.17" />
        <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="6.0.0-preview.5.21301.5" />
    </ItemGroup>

    <!-- Conditionally obtain references for the .NET Framework 5.0 target -->
    <ItemGroup Condition=" '$(TargetFramework)' == 'net5.0' ">
        <PackageReference Include="Microsoft.AspNetCore.Components" Version="5.0.7" />
        <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="5.0.7" />
        <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="5.0.0" />
    </ItemGroup>

    <!-- Workaround for this bug (https://github.com/dotnet/sdk/issues/18148) (replace the analyzer name with the one you need to exclude (filename only, no extension) -->
    <Target Name="RemoveLoggingAnalyzer" BeforeTargets="CoreCompile">
        <ItemGroup>
            <Analyzer Remove="@(Analyzer)" Condition="%(FileName) == 'Microsoft.Extensions.Logging.Generators'" />
        </ItemGroup>
    </Target>
```

**Comment by rainersigwald** [comment]
@MarkStega that error is being thrown in the `RazorCompileComponentDeclaration` target, which is before `CoreCompile`. Can you try changing your target to `AfterTargets="ResolvePackageAssets"`?

**Comment by MarkStega** [comment]
@rainersigwald No joy -- Here is a new log
[_Debug_AnyCPU_Build_2021-07-15T10_37_35.4829114-04_00.binlog.zip](https://github.com/dotnet/sdk/files/6823909/_Debug_AnyCPU_Build_2021-07-15T10_37_35.4829114-04_00.binlog.zip)
```
    <!-- Workaround for this bug (https://github.com/dotnet/sdk/issues/18148) (replace the analyzer name with the one you need to exclude (filename only, no extension) -->
    <Target Name="RemoveLoggingAnalyzer" AfterTargets="ResolvePackageAssets">
        <ItemGroup>
            <Analyzer Remove="@(Analyzer)" Condition="%(FileName) == 'Microsoft.Extensions.Logging.Generators'" />
        </ItemGroup>
    </Target>

**Comment by rainersigwald** [comment]
Sorry, try `AfterTargets="ResolveLockFileAnalyzers"`. That worked in my repro.

**Comment by MarkStega** [comment]
@rainersigwald Still not there
[_Debug_AnyCPU_Build_2021-07-15T11_05_53.6359804-04_00.binlog.zip](https://github.com/dotnet/sdk/files/6824161/_Debug_AnyCPU_Build_2021-07-15T11_05_53.6359804-04_00.binlog.zip)
